### PR TITLE
Create model

### PR DESCRIPTION
**Pull request by saleml**
Basic neural architecture.
Flexible params (to be defined via argparse in main script).

### CODE REVIEW

**Comment by rimassouel** [comment]
We should also add the discrete option to the latent variable distribution -> Multinouilli used for GMM

**Comment by rimassouel** [comment]
And enable the RP trick in the encode function -> if RP then sample = mean + eps*torch.exp(log sigma)

File: main.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+'''Main file to be called to train a model with a given algorithm'''
+
+from rws.model import BasicModel
+from argparse import ArgumentParser
+
+
```
Comment (lavoiems) [inline main.py:8]: Its good practive to guard the code with if __name__ == '__main__'

File: main.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+# Transform data
+# TODO: e.g. if MNIST, make sure input is transformed to 1d
+input_dim = None  # TODO: define this
+
+# Create model
+model = BasicModel(input_dim, args.hidden_dim, args.hidden_layers, args.encoding_dim,
+                   args.hidden_nonlinearity, args.decoder_nonlinearity)
+
+# Train model
+if args.algo == 'rws':
```
Comment (lavoiems) [inline main.py]: I guess we can do it this way, but we have to keep in mind that each models might have different hyper parameter.

File: rws/model.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
+import torch
+import torch.nn as nn
+from torch.distributions import Normal
+
+ACTIVATION_FUNCTIONS = {'tanh': nn.Tanh(),
```
Comment (lavoiems) [inline rws/model.py:5]: We should use inplace=True

File: rws/model.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
+import torch
```
Comment (lavoiems) [inline rws/model.py:1]: This should be in a different module, since it will be used by all the models

File: rws/model.py (hunk 3 above)
Comment (saleml) [reply rws/model.py:5]: not really. for example, I am using `out = self.encoder(input)`

File: main.py (hunk 1 above)
Comment (saleml) [reply main.py:8]: I don't understand this one

Comment (lavoiems) [reply main.py:8]: Official doc: https://docs.python.org/3/library/__main__.html

File: rws/model.py (hunk 3 above)
Comment (lavoiems) [reply rws/model.py:5]: I guess the comment should have been at the line of the relu. ReLU can be performed inplace. This has the advantage of saving memory.

File: main.py (hunk 1 above)
Comment (lavoiems) [reply main.py:8]: Maybe I can be more explicit. When you import a python module, all the code is being run, unless it's guarded. When a piece of code is guarded, it is only ran if the condition is met.

My comment is related to the code that is unguarded (i.e. all the lines below). Theses lines of code would be imported if we imported this file (e.g. if we ran unit test or wanted to load this module in an interpreter). It is possible to guard the code using the technique described in the doc that I linked in the comment above.

**Review by lavoiems — APPROVED** [review]
